src/Compilers/CSharp/Portable/Symbols/Extensions/SynthesizedExtensionMarker.cs

```diff
@@ -78,6 +78,29 @@ protected override (TypeWithAnnotations ReturnType, ImmutableArray<ParameterSymb
                 if (parameter is { })
                 {
                     checkUnderspecifiedGenericExtension(parameter, ContainingType.TypeParameters, diagnostics);
+
+                    TypeSymbol parameterType = parameter.TypeWithAnnotations.Type;
+                    RefKind parameterRefKind = parameter.RefKind;
+                    SyntaxNode? parameterTypeSyntax = parameterList.Parameters[0].Type;
+                    Debug.Assert(parameterTypeSyntax is not null);
+
+                    if (!parameterType.IsValidExtensionParameterType())
+                    {
+                        diagnostics.Add(ErrorCode.ERR_BadTypeforThis, parameterTypeSyntax, parameterType);
+                    }
+                    else if (parameterRefKind == RefKind.Ref && !parameterType.IsValueType)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_RefExtensionParameterMustBeValueTypeOrConstrainedToOne, parameterTypeSyntax);
+                    }
+                    else if (parameterRefKind is RefKind.In or RefKind.RefReadOnlyParameter && parameterType.TypeKind != TypeKind.Struct)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_InExtensionParameterMustBeValueType, parameterTypeSyntax);
+                    }
```

Comment:
I think it would be desirable to try sharing logic with ```ExtensionMethodChecks``` in order to avoid an accidental diverging

Reply:
I agree and I tried that initially, but the logic for these three checks in `ExtensionMethodChecks` uses locations and also the name of the method

Reply:
Consider leaving a comment in both places saying that we should keep them in sync.